cql, vector_search: implement read path

QuerthDP opened the pull request:
This pull request is an addition of ANN OF queries.

The patch contains:

- CQL syntax for ORDER BY `vector_column_name` ANN OF `vector_literal` clause of SELECT statements.
- implementation of external ANN queries (using vector-store service)
- tests

Example syntax:

```
SELECT comment
    FROM cycling.comments_vs
    ORDER BY comment_vector ANN OF [0.1, 0.15, 0.3, 0.12, 0.05]
    LIMIT 3;
```
Limit can be between 1 and 1000 - same as for Cassandra.

Co-authored-by: @janpiotrlakomy @smoczy123 
Fixes: VECTOR-48
Fixes: VECTOR-46

QuerthDP commented:
Waiting for https://github.com/scylladb/scylladb/pull/24331 to be merged until we start the review process.

QuerthDP commented:
The License Header Check will fail,
```
❌ Missing license header in: test/cqlpy/cassandra_tests/vector_invalid_query_test.py
```
as this test was translated from Cassandra, using their license:

```
# The original Apache Cassandra license:
#
# SPDX-License-Identifier: Apache-2.0
```

QuerthDP commented:
I believe we might start to review it.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17810/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17810/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17810/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/vector_store_test
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5626/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8811/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16895/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17810/)
#### Build Details:
- Duration: 7 hr 7 min
- Builder: spider2.cloudius-systems.com

QuerthDP commented:
@piodul - just a reminder, could you review this soon? I believe we're up to finally merge the https://github.com/scylladb/scylladb/pull/24331 and it would be appreciated to merge this quickly as well.

QuerthDP commented:
Changelog:

- rebased on master

QuerthDP commented on cql3/Cql.g:
```diff
@@ -484,11 +485,24 @@ whereClause returns [uexpression clause]
+                throw exceptions::invalid_request_exception(
+                    "Descending ANN ordering is not supported");
+            }
+            ann_orderings.emplace_back(c, ann_ordering.value());
```
What do you mean by "saving a pointer to a temporary object"? What exactly do you see wrong here?

As for the column_identifier - it's done exactly the same as in standard ordering.
As for the ann_ordering - getting the term, moving it to the ann_ordering optional, then pushing the existing value to the vector of expressions (this case pairs of column definition and expression) feels a proper way to do it imo. 
Very similar to handling collection literals for example.

nyh commented on cql3/statements/raw/select_statement.hh:
```diff
@@ -70,11 +74,13 @@ public:
     using compare_fn = std::function<bool(const T&, const T&)>;
 
     using result_row_type = std::vector<managed_bytes_opt>;
+    using prepared_ann_ordering_type = std::pair<const column_definition*, expr::expression>;
```
Since this is only a type, it's hard for me to come to any conclusions on how it is used, but I indeed see that the existing prepared_orderings_type uses exactly the same trick, so I have to assume it's safe.

smoczy123 commented on cql3/statements/select_statement.cc:
```diff
@@ -1145,7 +1183,41 @@ indexed_table_select_statement::do_execute(query_processor& qp,
+        auto pkeys = co_await qp.vector_store_client().ann(_schema->ks_name(), _index.metadata().name(), _schema , std::move(ann_vector), limit, as);
+
+        if (!pkeys.has_value()) {
+            co_await coroutine::return_exception(exceptions::invalid_request_exception(fmt::format("ANN query cannot be executed with vector_store disabled")));
```
Please consider forwarding the message of the error returned by the vector store instead.

nyh commented on cql3/statements/select_statement.cc:
```diff
@@ -2371,6 +2430,57 @@ void select_statement::verify_ordering_is_valid(const prepared_orderings_type& o
+    }
+
+    if (!def->type->is_vector() || static_cast<const vector_type_impl*>(def->type.get())->get_elements_type()->get_kind() != abstract_type::kind::float_kind) {
+        throw exceptions::invalid_request_exception("ANN ordering is only supported on float vector indexes");
```
Is this limitation ("quantization" not supported!) shared by Cassandra?

QuerthDP replied:
In Cassandra: https://github.com/apache/cassandra/blob/53eb2fdac886fb559ce36649d97c3a87a0423e58/src/java/org/apache/cassandra/cql3/restrictions/StatementRestrictions.java#L303-L304

It is done exactly the same way, with the exact message.

nyh commented on db/schema_tables.cc:
```diff
@@ -1840,7 +1840,7 @@ void add_table_or_view_to_schema_mutation(schema_ptr s, api::timestamp_type time
 static void make_drop_table_or_view_mutations(schema_ptr schema_table, schema_ptr table_or_view, api::timestamp_type timestamp, utils::chunked_vector<mutation>& mutations);
 
-static bool should_create_view(const index_metadata & index) {
+bool should_create_view(const index_metadata & index) {
```
As I noted above, perhaps if you start to use this function more widely, its name should be changed.

QuerthDP commented on test/cqlpy/cassandra_tests/vector_invalid_query_test.py:
```diff
@@ -52,93 +50,99 @@
+def test_cannot_query_wrong_number_of_dimensions(cql, test_keyspace):
+    with create_table(cql, test_keyspace, "(pk int, str_val text, val vector<float, 3>, PRIMARY KEY(pk))") as table:
+        if is_scylla(cql):
+            execute(cql, table, "CREATE CUSTOM INDEX ON %s(val) USING 'vector_index'")
```
Good point, will change that. Looks more tidy.

ewienik commented:
> > Do we have a way to do functional testing on the last patch - i.e., that searching with a vector store actually works correctly, and works correctly both if we do have clustering key and if we don't (because you have different code paths for that).
> 
> @ewienik is working on it. It will not be included in this PR, as the logic upon integration tests with Vector Store is complex and require lots of work and preparation.
> 
> I believe this refers to the case: [Vector Search Test Plan](https://scylladb.atlassian.net/wiki/spaces/RND/pages/39518215/Vector+Search+Test+Plan+Document)

Infrastructure for integration tests are in development under https://scylladb.atlassian.net/browse/VECTOR-52
In the Vector Search Test Plan there will be an overall description how this integration/functional tests should look like.

QuerthDP commented on test/cqlpy/cassandra_tests/vector_invalid_query_test.py:
```diff
@@ -52,93 +50,99 @@
+    with create_table(cql, test_keyspace, "(pk int, str_val text, val1 vector<float, 3>, val2 vector<float, 3>, PRIMARY KEY(pk))") as table:
+
+        # Scylla doesnt support custom indexes on text columns so we use a regular index.
+        execute(cql, table, "CREATE INDEX ON %s(str_val)")
```
I looked at it now and I laughed. Looks silly indeed.
I believe @janpiotrlakomy took the strategy to translate all possible tests that work with our implementation, and that's why it's here.
But I think it's Cassandra to blame for having such test.

QuerthDP commented on test/cqlpy/cassandra_tests/vector_invalid_query_test.py:
```diff
@@ -52,93 +50,99 @@
+            "SELECT * FROM %s ORDER BY val ann of [2.5, 3.5, 4.5] LIMIT 5"
+        )
+        assert_invalid_message(
+            cql, table, ANN_REQUIRES_INDEXED_FILTERING_MESSAGE if is_scylla(cql) else ANN_REQUIRES_INDEX_MESSAGE,
```
Also I have no idea why such a misleading message is being shown.
We check if filtering is on, and then we tell the user that it requires all restricted (does not that mean vector indexed in our case?) columns to be indexed. 
I believe something is messed up here, and that was not Cassandra's intention.
Will take a look on it. I think more testing is needed.

QuerthDP replied:
Finally I've changed the message to state that we do not allow filtering with ANN queries at all (we don't have it implemented as it's unclear what approach should we take).

It looks like Cassandra have some errors here as well:
When you follow their error messages and create all needed indexes - the query seems to work.
But when you add some data and try the query once again the assertion error is thrown 🤷

QuerthDP commented:
FYI - the CI will fail, there are some tests failing yet to be fixed

TODO: 
- fix issues with `WHERE` clause and `ALLOW FILTERING`

Changelog:
- fixed nitpicks
- made the vector index options case insensitive
- normalized the Cassandra tests to differ only by `vector_index` vs `StorageAttachedIndex`
- introduced more informative errors if ANN queries fail
- unified some error messages with Cassandra ones (easier to test)

QuerthDP commented on cql3/restrictions/statement_restrictions.cc:
```diff
@@ -1274,14 +1275,16 @@ statement_restrictions::statement_restrictions(data_dictionary::database db,
-        sstring index_table_name = im.name() + "_index";
-        schema_ptr view_schema = db.find_schema(schema->ks_name(), index_table_name);
-        _view_schema = view_schema;
+        if (db::schema_tables::should_create_view(im)) {
```
Changing to `view_should_exist`. I feel like the name is accurate both for checking if the view should be created and if the view has been created.

QuerthDP commented on cql3/statements/raw/select_statement.hh:
```diff
@@ -46,19 +46,23 @@ public:
     class parameters final {
     public:
         using orderings_type = std::vector<std::pair<shared_ptr<column_identifier::raw>, ordering>>;
+        using ann_orderings_type = std::vector<std::pair<shared_ptr<column_identifier::raw>, expr::expression>>;
```
> In the syntax I see "ANN OF [0.1, 0.15, 0.3, 0.12, 0.05]" - where is the expression here, the vector? So it's basically an expression that is expected evaluate to a vector of the right length? Or what?

It requires it to be an expression that should be evaluated to vector of floats with expected length.

QuerthDP replied:
The checks are done in `prepare_ann_ordering` -> `prepare_expression`

QuerthDP replied:
> I had a thought - do we really need another list of order parameters?

We did not. Changed that to one list with variant.

QuerthDP commented:
Changelog: 

- resolved remaining comments
- fixed failing tests
- disallowed all filtering when performing ANN queries
- renamed `should_create_view` to `view_should_exist`

QuerthDP commented:
There seems to be an CI issue that leads to failing CI actions. @scylladb/releng?

QuerthDP commented:
> Sorry, that was me doing something in the servers (updating with ansible playbook), re-kicked the CI

It still failed with the same errors.

QuerthDP commented:
> #### 🔴 CI State: FAILURE
> #### Failed Tests (1/73225):
> * [test_system_clients](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/18713/testReport/junit/cql_system_clients_test/TestSystemClients/release___dtest_with_consistent_topology_changes___test_system_clients) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_system_clients)

It is a known issue ( https://github.com/scylladb/scylladb/issues/25253), rerunning the CI.

knowack1 commented on cql3/statements/select_statement.cc:
```diff
@@ -994,6 +1010,30 @@ indexed_table_select_statement::prepare(data_dictionary::database db,
     auto& sim = cf.get_index_manager();
     auto [index_opt, used_index_restrictions] = restrictions->find_idx(sim);
+
+    if (prepared_ann_ordering.has_value()) {
```
Consider to created dedicated select_statement implementation instead of adding implementation to the indexed_table_select_statement

QuerthDP replied:
I think it might be nice to to someday. But will leave it for now, as it's urgent to merge this PR.
Please create a jira ticket for it.

QuerthDP replied:
https://scylladb.atlassian.net/browse/VECTOR-162

QuerthDP commented:
Changelog:

- fixed a typo
- moved conversion from data_value to float to a function

nyh commented:
Note for posterity: The "license check" that fails is a false-positive. It fails for test/cqlpy/cassandra_tests/vector_invalid_query_test.py. This is a translated Cassandra test, so we kept the original Cassandra copyright and license. Under US copyright law (at least), to add our own copyright some level of creativity must be involved in the translation - a "mindless" literal translation from Java to Python doesn't quite cut it. Let's not add our copyright to this file. We didn't add it to any of the other files in test/cqlpy/cassandra_tests.

avikivity commented:
Please follow up with reference documentation updates.
